### PR TITLE
ipq806x: add support for Nokia Airscale AC400i

### PR DESCRIPTION
**Pull request by krusic22**
Adds support for the Nokia Airscale WI4A-AC400i Access Point.
The device doesn't really exist outside of public infrastructure,
so no public page about it exist.
But can be looked up using the FCC ID: [2AD8UFZCWI4A1](https://fccid.io/2AD8UFZCWI4A1)

Hardware:
```
SoC:    Qualcomm IPQ8065
RAM:    512 MB DDR3
Flash:  256 MB NAND (Macronix MX30UF2G18AC)
        4 MB SPI-NOR (Macronix MX25U3235F)
WLAN:   Qualcomm Atheros QCA9984 - 2.4Ghz
        Qualcomm Atheros QCA9984 - 5Ghz
ETH:    eth0 - POE (100Mbps in U-Boot, 1000Mbps in OpenWrt)
        eth1 - (1000Mbps in both)
        Auto-negotiation broken on both.
USB:    USB 2.0
LED:    5G, 2.4G, ETH1, ETH2, CTRL, PWR (All support green and red)
BTN:    Reset
Other:  SD card slot (non-functional)
Serial: 115200bps, near the Ethernet transformers, labeled 9X.
        Connections from the arrow to the 9X text:
		[NC] - [TXD] - [GND] - [RXD] - [NC]
```

Install instructions are in the commit description and on the [forum](https://forum.openwrt.org/t/adding-openwrt-support-for-nokia-airscale-ac400i/78356/59?u=2in2). 

Notes:
- Auto-negotiation is broken in U-Boot and OEM firmware, so both ports are forced to run at 1000Mbps.
The POE port runs at 100Mbps in U-boot.
- The SD card slot is also non-functional (just like in the OEM firmware).
- The NAND chip is split into two partitions inside U-Boot (PartA and PartB),
we are installing OpenWrt onto PartA and leaving PartB alone, so restoring
the OEM firmware, is as easy switching the boot partition.

### CODE REVIEW

File: target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 1]
```diff
@@ -0,0 +1,338 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq8065.dtsi"
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:3]: this should use the smb variant

File: target/linux/ipq806x/image/generic.mk [hunk 2]
```diff
@@ -337,6 +337,20 @@ define Device/netgear_xr500
...
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Nokia
+	DEVICE_MODEL := AC400i
+	SOC := qcom-ipq8065
+	DEVICE_DTS := qcom-ipq8065-ac400i
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := ac400i
+	DEVICE_PACKAGES := -swconfig ath10k-firmware-qca9984-ct ipq-wifi-nokia-ac400i
```
Comment (Ansuel) [inline target/linux/ipq806x/image/generic.mk:350]: Does having the swconfig package a problem? I would keep it for consistency.

**Comment by Ansuel** [comment]
Honestly I would drop the 5.10 changes since 5.15 is currently the default.

File: target/linux/ipq806x/image/generic.mk (hunk 2 above)
Comment (krusic22) [reply target/linux/ipq806x/image/generic.mk]: Looks like it's not required anymore. So, removed.

File: target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts (hunk 1 above)
Comment (krusic22) [reply target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts]: Marking as resolved, since I dropped 5.10 changes.

Comment (Ansuel) [reply target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts]: this also apply to 5.15

Comment (krusic22) [reply target/linux/ipq806x/files-5.10/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts]: Done.

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+		nss@40000000 {
+			reg = <0x40000000 0x1000000>;
+			no-map;
+		};
+
+		smem: smem@41000000 {
+			reg = <0x41000000 0x200000>;
+			no-map;
+		};
+
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:20]:
```suggestion
```

Please drop these already defined in ipq8064 dtsi

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+};
+
+&nand {
+	status = "okay";
+	pinctrl-0 = <&nand_pins>;
+	pinctrl-names = "default";
+
+	nand@0 {
+		reg = <0>;
+		compatible = "qcom,nandcs";
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:299]: add an extra space after this

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+		nand-ecc-strength = <4>;
+		nand-bus-width = <8>;
+		nand-ecc-step-size = <512>;
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:302]: add an extra space after this

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:306]: add an extra space after this

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+			reg = <0x40000000 0x1000000>;
+			no-map;
+		};
+
+		smem: smem@41000000 {
+			reg = <0x41000000 0x200000>;
+			no-map;
+		};
+
+		wifi_dump@44000000 {
```
Comment (Ansuel) [inline target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts:21]: Is this really needed and used?

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts (hunk 3 above)
Comment (krusic22) [reply target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts]: Removed.

File: target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts (hunk 7 above)
Comment (krusic22) [reply target/linux/ipq806x/files-5.15/arch/arm/boot/dts/qcom-ipq8065-ac400i.dts]: Looks like it's no longer needed.

**Comment by Ansuel** [comment]
@krusic22 we are moving the wifi board file to a separate repo... I would wait for that change to be done and then i will work on merging this. Sorry for the delay.

**Comment by Ansuel** [comment]
@krusic22 the repo is now present. Please submit the BDF there with a pr.

**Comment by krusic22** [comment]
> @krusic22 the repo is now present. Please submit the BDF there with a pr.

Done: https://github.com/openwrt/firmware_qca-wireless/pull/2

**Comment by Ansuel** [comment]
@krusic22 you added the variant naming or they were like that in the original fw? if so I would drop the -2G -5G name from them

**Comment by krusic22** [comment]
I added the suffix to differentiate between the 2 files.

**Comment by Ansuel** [comment]
@krusic22 hi it's not needed since they will use the board-id so please drop them and keep the same variant. Do you want me to do the change or you can do it?

**Comment by krusic22** [comment]
You may go ahead and change it. Thank you.

**Comment by Ansuel** [comment]
@krusic22 ok now you just need to bump ipq-wifi (include also you changes to the makefile) in a separate commit, give this a final test to make sure the BDF work and you are good to go

**Comment by krusic22** [comment]
Do I have to change anything?
I just applied the patch/PR over the latest master, used the master/ipq806x config.buildinfo,
choose the AP in the menu config and that's it.
Everything works as expected.

**Comment by Ansuel** [comment]
@krusic22 the calibratio needs to be fixed... and as requested the ipq-wifi changes needs to be put in a separate commit. Check the prev commit of that package to understand the format of an ipq-wifi bump commit

**Comment by krusic22** [comment]
> @krusic22 the calibratio needs to be fixed... and as requested the ipq-wifi changes needs to be put in a separate commit. Check the prev commit of that package to understand the format of an ipq-wifi bump commit

Done.
_I hope at least._

**Pull request merged**